File: apps/host/phantom_host/wand_protocol.py
"""Wire codec for BADGE-FIRMWARE-CONTRACT.md v1 (the badge's GATT records).

Every record is exactly 20 bytes, little-endian, version 1. This module mirrors the badge
firmware (firmware/src/proto.cpp) and the browser decoder (apps/web/src/wand/protocol.ts);
``tests/test_wand_protocol.py`` pins all three to the contract's golden vectors.

The referee never talks to a badge directly (the browser does), so this module exists for
tooling, traces and tests. Decoding never raises for wire-level problems: ``None`` means the
record is malformed and must be ignored.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass

VERSION = 1
RECORD_LEN = 20

# opcodes
OP_OPEN, OP_SYNC, OP_SET_STATE, OP_CUE = 1, 2, 3, 4
# result codes
R_OK, R_MALFORMED, R_WRONG_SESSION, R_INVALID_ARG, R_EXPIRED, R_UNSUPPORTED, R_STALE_SEQ = range(7)
# phases
PH_IDLE, PH_PRACTICE, PH_COUNTDOWN, PH_PLAYING, PH_WON, PH_LOST, PH_DRAW, PH_ABORTED = range(8)
# effects / spells
FX_ACCEPTED_CAST, FX_BLOCKED, FX_DAMAGE, FX_RESULT = 1, 2, 3, 4
SP_NONE, SP_STUPEFY, SP_PROTEGO, SP_EXPELLIARMUS = 0, 1, 2, 3
# flags / bits
MF_VALID, MF_SATURATED, MF_DISCONTINUITY = 1, 2, 4
H_SENSOR, H_STREAM, H_PRESENTATION, H_STATE_STALE = 1, 2, 4, 8
CAP_ALL = 0x0F
RANGE_MG = 8000


@dataclass(frozen=True)
class Info:
    caps: int
    sample_hz: int
    range_g: int
    device_id: bytes  # 6 bytes
    boot_id: int
    fw: tuple[int, int, int]
    axis_convention: int


@dataclass(frozen=True)
class Motion:
    flags: int
    seq: int
    capture_ms: int
    boot_id: int
    ax: int
    ay: int
    az: int

    @property
    def valid(self) -> bool:
        return bool(self.flags & MF_VALID)

    @property
    def saturated(self) -> bool:
        return bool(self.flags & MF_SATURATED)

    @property
    def discontinuity(self) -> bool:
        return bool(self.flags & MF_DISCONTINUITY)


@dataclass(frozen=True)
class Control:
    opcode: int
    seq: int
    nonce: int
    arg0: int = 0
    arg1: int = 0
    arg2: int = 0


@dataclass(frozen=True)
class Status:
    kind: int  # 0 health, 1 command result
    seq: int
    nonce: int
    device_ms: int
    detail0: int
    detail1: int


def encode_info(v: Info) -> bytes:
    if len(v.device_id) != 6:
        raise ValueError("device_id must be 6 bytes")
    return struct.pack("<BBBB6sIBBBBH", VERSION, v.caps, v.sample_hz, v.range_g, v.device_id, v.boot_id, *v.fw, v.axis_convention, 0)


def decode_info(data: bytes) -> Info | None:
    if len(data) != RECORD_LEN or data[0] != VERSION or data[18:20] != b"\x00\x00":
        return None
    _, caps, hz, rg, dev, boot, a, b, c, axis, _ = struct.unpack("<BBBB6sIBBBBH", data)
    return Info(caps, hz, rg, dev, boot, (a, b, c), axis)


def encode_motion(v: Motion) -> bytes:
    return struct.pack("<BBHIIhhhH", VERSION, v.flags, v.seq & 0xFFFF, v.capture_ms & 0xFFFFFFFF, v.boot_id & 0xFFFFFFFF, v.ax, v.ay, v.az, 0)


def decode_motion(data: bytes) -> Motion | None:
    if len(data) != RECORD_LEN or data[0] != VERSION:
        return None
    _, flags, seq, capture, boot, ax, ay, az, reserved = struct.unpack("<BBHIIhhhH", data)
    if reserved or flags & ~(MF_VALID | MF_SATURATED | MF_DISCONTINUITY):
        return None
    if any(abs(a) > RANGE_MG for a in (ax, ay, az)):
        return None
    return Motion(flags, seq, capture, boot, ax, ay, az)


def encode_control(v: Control) -> bytes:
    return struct.pack("<BBHIIII", VERSION, v.opcode, v.seq & 0xFFFF, v.nonce & 0xFFFFFFFF, v.arg0 & 0xFFFFFFFF, v.arg1 & 0xFFFFFFFF, v.arg2 & 0xFFFFFFFF)


def decode_control(data: bytes) -> Control | None:
    if len(data) != RECORD_LEN or data[0] != VERSION:
        return None
    _, op, seq, nonce, a0, a1, a2 = struct.unpack("<BBHIIII", data)
    return Control(op, seq, nonce, a0, a1, a2)


def encode_status(v: Status) -> bytes:
    return struct.pack("<BBHIIII", VERSION, v.kind, v.seq & 0xFFFF, v.nonce & 0xFFFFFFFF, v.device_ms & 0xFFFFFFFF, v.detail0 & 0xFFFFFFFF, v.detail1 & 0xFFFFFFFF)


def decode_status(data: bytes) -> Status | None:
    if len(data) != RECORD_LEN or data[0] != VERSION:
        return None
    _, kind, seq, nonce, ms, d0, d1 = struct.unpack("<BBHIIII", data)
    return Status(kind, seq, nonce, ms, d0, d1)


def set_state_args(phase: int, hp: int, status: int = 0, maxhp: int = 100) -> int:
    """Pack SET_STATE arg0 (least significant byte first: phase, hp, maxhp, status)."""
    return phase | (hp << 8) | (maxhp << 16) | (status << 24)


def cue_args(effect: int, spell: int, duration_ms: int) -> int:
    return effect | (spell << 8) | (duration_ms << 16)


def seq_advances(new: int, old: int) -> bool:
    """Contract rule for 16-bit sequences: accept only 0 < (new - old) mod 65536 < 32768."""
    d = (new - old) % 65536
    return 0 < d < 32768

File: apps/web/src/lib/wandProtocol.ts
// Wire codec for BADGE-FIRMWARE-CONTRACT.md v1: the badge's GATT records.
// Every record is exactly 20 bytes, little-endian, version 1. Mirrors firmware/src/proto.cpp and
// apps/host/phantom_host/wand_protocol.py; wandProtocol.test.ts pins all three to the golden vectors.
// Decoders return null for anything malformed: a bad record is ignored, never guessed at.

export const VERSION = 1;
export const RECORD_LEN = 20;
export const RANGE_MG = 8000;

export const SERVICE_UUID = "7f510000-1b15-4f0d-8f3c-8db47a812000";
export const INFO_UUID = "7f510001-1b15-4f0d-8f3c-8db47a812000";
export const MOTION_UUID = "7f510002-1b15-4f0d-8f3c-8db47a812000";
export const CONTROL_UUID = "7f510003-1b15-4f0d-8f3c-8db47a812000";
export const STATUS_UUID = "7f510004-1b15-4f0d-8f3c-8db47a812000";

export const OP = { OPEN: 1, SYNC: 2, SET_STATE: 3, CUE: 4 } as const;
export const RC = { OK: 0, MALFORMED: 1, WRONG_SESSION: 2, INVALID_ARG: 3, EXPIRED: 4, UNSUPPORTED: 5, STALE_SEQ: 6 } as const;
export const RC_NAMES = ["ok", "malformed", "wrong session", "invalid argument", "expired", "unsupported", "stale sequence"];
export const PHASE = { IDLE: 0, PRACTICE: 1, COUNTDOWN: 2, PLAYING: 3, WON: 4, LOST: 5, DRAW: 6, ABORTED: 7 } as const;
export const EFFECT = { ACCEPTED_CAST: 1, BLOCKED: 2, DAMAGE: 3, RESULT: 4 } as const;
export const SPELL = { NONE: 0, STUPEFY: 1, PROTEGO: 2, EXPELLIARMUS: 3 } as const;
export const STATUS_BIT = { SHIELD: 1, LOCKED: 2 } as const;
export const MOTION_FLAG = { VALID: 1, SATURATED: 2, DISCONTINUITY: 4 } as const;
export const HEALTH = { SENSOR: 1, STREAM: 2, PRESENTATION: 4, STATE_STALE: 8 } as const;
export const CAP_ALL = 0x0f;

export type Info = {
  caps: number;
  sampleHz: number;
  rangeG: number;
  deviceId: Uint8Array; // 6 bytes
  bootId: number;
  fw: [number, number, number];
  axisConvention: number;
};

export type Motion = {
  flags: number;
  seq: number;
  captureMs: number;
  bootId: number;
  ax: number;
  ay: number;
  az: number;
};

export type Control = { opcode: number; seq: number; nonce: number; arg0: number; arg1: number; arg2: number };
export type Status = { kind: number; seq: number; nonce: number; deviceMs: number; detail0: number; detail1: number };

function view(buf: Uint8Array): DataView {
  return new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
}

export function encodeInfo(v: Info): Uint8Array {
  const out = new Uint8Array(RECORD_LEN);
  const d = view(out);
  out[0] = VERSION;
  out[1] = v.caps;
  out[2] = v.sampleHz;
  out[3] = v.rangeG;
  out.set(v.deviceId.subarray(0, 6), 4);
  d.setUint32(10, v.bootId >>> 0, true);
  out[14] = v.fw[0];
  out[15] = v.fw[1];
  out[16] = v.fw[2];
  out[17] = v.axisConvention;
  return out;
}

export function decodeInfo(buf: Uint8Array): Info | null {
  if (buf.length !== RECORD_LEN || buf[0] !== VERSION || buf[18] !== 0 || buf[19] !== 0) return null;
  const d = view(buf);
  return {
    caps: buf[1],
    sampleHz: buf[2],
    rangeG: buf[3],
    deviceId: buf.slice(4, 10),
    bootId: d.getUint32(10, true),
    fw: [buf[14], buf[15], buf[16]],
    axisConvention: buf[17],
  };
}

export function encodeMotion(v: Motion): Uint8Array {
  const out = new Uint8Array(RECORD_LEN);
  const d = view(out);
  out[0] = VERSION;
  out[1] = v.flags;
  d.setUint16(2, v.seq & 0xffff, true);
  d.setUint32(4, v.captureMs >>> 0, true);
  d.setUint32(8, v.bootId >>> 0, true);
  d.setInt16(12, v.ax, true);
  d.setInt16(14, v.ay, true);
  d.setInt16(16, v.az, true);
  return out;
}

export function decodeMotion(buf: Uint8Array): Motion | null {
  if (buf.length !== RECORD_LEN || buf[0] !== VERSION || buf[18] !== 0 || buf[19] !== 0) return null;
  if (buf[1] & ~(MOTION_FLAG.VALID | MOTION_FLAG.SATURATED | MOTION_FLAG.DISCONTINUITY)) return null;
  const d = view(buf);
  const m: Motion = {
    flags: buf[1],
    seq: d.getUint16(2, true),
    captureMs: d.getUint32(4, true),
    bootId: d.getUint32(8, true),
    ax: d.getInt16(12, true),
    ay: d.getInt16(14, true),
    az: d.getInt16(16, true),
  };
  if (Math.abs(m.ax) > RANGE_MG || Math.abs(m.ay) > RANGE_MG || Math.abs(m.az) > RANGE_MG) return null;
  return m;
}

export function encodeControl(v: Control): Uint8Array {
  const out = new Uint8Array(RECORD_LEN);
  const d = view(out);
  out[0] = VERSION;
  out[1] = v.opcode;
  d.setUint16(2, v.seq & 0xffff, true);
  d.setUint32(4, v.nonce >>> 0, true);
  d.setUint32(8, v.arg0 >>> 0, true);
  d.setUint32(12, v.arg1 >>> 0, true);
  d.setUint32(16, v.arg2 >>> 0, true);
  return out;
}

export function decodeControl(buf: Uint8Array): Control | null {
  if (buf.length !== RECORD_LEN || buf[0] !== VERSION) return null;
  const d = view(buf);
  return { opcode: buf[1], seq: d.getUint16(2, true), nonce: d.getUint32(4, true), arg0: d.getUint32(8, true), arg1: d.getUint32(12, true), arg2: d.getUint32(16, true) };
}

export function encodeStatus(v: Status): Uint8Array {
  const out = new Uint8Array(RECORD_LEN);
  const d = view(out);
  out[0] = VERSION;
  out[1] = v.kind;
  d.setUint16(2, v.seq & 0xffff, true);
  d.setUint32(4, v.nonce >>> 0, true);
  d.setUint32(8, v.deviceMs >>> 0, true);
  d.setUint32(12, v.detail0 >>> 0, true);
  d.setUint32(16, v.detail1 >>> 0, true);
  return out;
}

export function decodeStatus(buf: Uint8Array): Status | null {
  if (buf.length !== RECORD_LEN || buf[0] !== VERSION) return null;
  const d = view(buf);
  return { kind: buf[1], seq: d.getUint16(2, true), nonce: d.getUint32(4, true), deviceMs: d.getUint32(8, true), detail0: d.getUint32(12, true), detail1: d.getUint32(16, true) };
}

/** SET_STATE arg0: phase, HP, maxHP, status, least significant byte first. */
export function setStateArgs(phase: number, hp: number, status = 0, maxhp = 100): number {
  return (phase | (hp << 8) | (maxhp << 16) | (status << 24)) >>> 0;
}

/** CUE arg0: effect, spell, duration in ms. */
export function cueArgs(effect: number, spell: number, durationMs: number): number {
  return (effect | (spell << 8) | (durationMs << 16)) >>> 0;
}

/** Contract rule for 16-bit sequences: accept only 0 < (new - old) mod 65536 < 32768. */
export function seqAdvances(next: number, prev: number): boolean {
  const d = (next - prev + 65536) % 65536;
  return d > 0 && d < 32768;
}

/** Wrap-safe difference of two 32-bit millisecond clocks (a - b). */
export function diff32(a: number, b: number): number {
  return ((a - b) | 0);
}

export function hex(buf: Uint8Array): string {
  return Array.from(buf, (b) => b.toString(16).padStart(2, "0")).join(" ");
}

export function fromHex(text: string): Uint8Array {
  const clean = text.replace(/[^0-9a-fA-F]/g, "");
  const out = new Uint8Array(clean.length / 2);
  for (let i = 0; i < out.length; i++) out[i] = parseInt(clean.slice(i * 2, i * 2 + 2), 16);
  return out;
}

export function deviceIdHex(id: Uint8Array): string {
  return Array.from(id, (b) => b.toString(16).toUpperCase().padStart(2, "0")).join("");
}
